sway/desktop/xdg_shell.c

```diff
@@ -87,32 +63,53 @@ static void popup_unconstrain(struct sway_xdg_popup *popup) {
 	wlr_xdg_popup_unconstrain_from_box(wlr_popup, &output_toplevel_sx_box);
 }
 
-static struct sway_xdg_popup *popup_create(
-		struct wlr_xdg_popup *wlr_popup, struct sway_view *view) {
+static struct sway_xdg_popup *popup_create(struct wlr_xdg_popup *wlr_popup,
+		struct sway_view *view, struct wlr_scene_tree *parent) {
 	struct wlr_xdg_surface *xdg_surface = wlr_popup->base;
 
-	struct sway_xdg_popup *popup =
-		calloc(1, sizeof(struct sway_xdg_popup));
-	if (popup == NULL) {
+	struct sway_xdg_popup *popup = calloc(1, sizeof(struct sway_xdg_popup));
+	if (!popup) {
+		return NULL;
+	}
+
+	popup->scene_tree = wlr_scene_tree_create(parent);
+	if (!popup->scene_tree) {
+		free(popup);
+		return NULL;
+	}
+
+	popup->xdg_surface_tree = wlr_scene_xdg_surface_create(
+		popup->scene_tree, xdg_surface);
+	if (!popup->xdg_surface_tree) {
+		wlr_scene_node_destroy(&popup->scene_tree->node);
```

Comment:
Do we need to free `popup->scene_tree`, or does `wlr_scene_node_destroy` have some magic to figure it out based off `->node`?

Reply:
It has magic.